### PR TITLE
fix(server): ensure distributed serving / serving on all platforms

### PR DESCRIPTION
**Pull request by bojiang**
## Description

* internal serve API
* add e2e tests to ensure it
* many fixes for the workflow on Windows

## Checklist:

- [ ] My code follows the bentoml code style, both `make format` and
  `make lint` script have passed
  ([instructions](https://github.com/bentoml/BentoML/blob/master/DEVELOPMENT.md#style-check-and-auto-formatting-your-code)).
- [ ] My change reduces project test coverage and requires unit tests to be added
- [ ] I have added unit tests covering my code change
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: bentoml/_internal/server/cli/runner.py [hunk 1]
```diff
@@ -38,8 +49,39 @@ def main(
...
+        from bentoml._internal.utils.click import unparse_click_params
+        from bentoml._internal.utils.circus import create_standalone_arbiter
+        from bentoml._internal.utils.circus import create_circus_socket_from_uri
+
+        ensure_prometheus_dir()
+        circus_socket = create_circus_socket_from_uri(bind, name=runner_name)
+        params = ctx.params
+        params["bind"] = f"fd://$(circus.sockets.{runner_name})"
+        params["as_worker"] = True
+        print(unparse_click_params(params, ctx.command.params))
```
Comment (aarnphm) [inline bentoml/_internal/server/cli/runner.py:69]:
```suggestion
```

File: .gitattributes [hunk 2]
```diff
@@ -1 +1,2 @@
 typings/**/*.pyi linguist-generated=true
+* text=auto eol=lf
```
Comment (bojiang) [inline .gitattributes:2]: avoid "\r\n" on Windows, which cause a fatal error in generated bentos

File: bentoml/_internal/utils/platform.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+
+def cancel_subprocess(p: "subprocess.Popen[t.Any]") -> None:
+    """
+    Tell the process to terminate and kill all of its children. Availabe both on Windows and Linux.
+    Note: It will return immediately rather than wait for the process to terminate.
+
+    Args:
+        p: subprocess.Popen object
+    """
```
Comment (bojiang) [inline bentoml/_internal/utils/platform.py:32]:
```suggestion
```

File: bentoml/testing/server.py [hunk 4]
```diff
@@ -176,12 +182,12 @@ def run_api_server_in_docker(
                     f"API server {host_url} failed to start within {timeout} seconds"
                 )
         finally:
-            proc.terminate()
+            subprocess.check_call(["docker", "stop", container_name])
```
Comment (bojiang) [inline bentoml/testing/server.py:185]:
```suggestion
            subprocess.call(["docker", "stop", container_name])
```

File: bentoml/testing/server.py [hunk 5]
```diff
@@ -193,6 +199,10 @@ def run_api_server(
     """
     workdir = workdir if workdir is not None else "./"
     my_env = os.environ.copy()
+    my_env["PYTHONIOENCODING"] = "utf-8"
```
Comment (bojiang) [inline bentoml/testing/server.py:202]:
```suggestion
```

File: bentoml/testing/server.py [hunk 6]
```diff
@@ -203,30 +213,117 @@ def run_api_server(
...
 
     logger.info(f"Running command: `{cmd}`")
-    if config_file is not None:
-        my_env["BENTOML_CONFIG"] = os.path.abspath(config_file)
 
     p = subprocess.Popen(
         cmd,
         stderr=subprocess.STDOUT,
         env=my_env,
+        encoding="utf-8",
```
Comment (bojiang) [inline bentoml/testing/server.py:221]:
```suggestion
```

File: bentoml/testing/server.py [hunk 7]
```diff
@@ -203,30 +213,117 @@ def run_api_server(
...
+                "--runner-name",
+                runner["name"],
+            ]
+
+            logger.info(f"Running command: `{cmd}`")
+
+        processes.append(
+            subprocess.Popen(
+                cmd,
+                encoding="utf-8",
```
Comment (bojiang) [inline bentoml/testing/server.py:281]:
```suggestion
```

File: bentoml/testing/server.py [hunk 8]
```diff
@@ -203,30 +213,117 @@ def run_api_server(
...
+            "--working-dir",
+            path,
+        ]
+        logger.info(f"Running command: `{cmd}`")
+
+    processes.append(
+        subprocess.Popen(
+            cmd,
+            stderr=subprocess.STDOUT,
+            encoding="utf-8",
```
Comment (bojiang) [inline bentoml/testing/server.py:306]:
```suggestion
```

**Pull request merged**